DOSE-475 Zpool iostat

don-brady opened the pull request:
### Motivation and Context
Add zpool iostat stats option for object based pools. 

### Description
The new "-o" option will display stats broke out by categories -- demand reads and txg sync writes (aka object_store_data), metadata, reclaims (background frees), and object deletions.  For each GET/PUT it displays the operations and throughput values.

In addition, the "-ol" option will add latencies for I/O roundtrip to object agent and the latencies for GET/PUT requests (the latter is still a WIP).

Also fixed the stock "-w" and "-l" displays when the pool is object based to not have bogus values (since the object store VDEV doesn't use the vdev_queue layer).

### Implementation Details
The `zpool` command still uses an ioctl to get the stats from the kernel.  For an object based pool, the kernel will now make a request to the object agent for the stats for that pool.  On the agent side, the stats are collected and stored in `ObjectAccess`. The `get_object` and `put_object` methods now take a `stat_type` parameter so the proper stat can be updated.

### Open Issues
1. ~The bytes value for the `ObjectDelete` stat is not yet supplied from the agent.~ Note - byte column was removed.
2. The kernel will collect the stats from the agent for operations other than `zpool iostat -o`
3. ~During development I encountered what appeared to be a lost response from the agent and the thread waiting for the response hangs in cv_wait().~
4. ~With moderate load I can see the iostat display miss reporting at a one second interval (stalls in kernel).~

### How Has This Been Tested?
Manual runs of `zpool iostat -o [-l]`

### Sample Output
**`zpool iostat -o <pool>`**
```
                                object_agent_io            object_store_data         object_store_metadata       object_store_reclaim
              capacity     operations    throughput    operations    throughput    operations    throughput    operations    throughput     deletions
pool        alloc   free   read  write   read  write    get    put    get    put    get    put    get    put    get    put    get    put    ops   rate
----------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
delta       15.2G   512P    140  21.9K   328K  22.7M     15     21  14.8M  20.6M      0      2  1.62K  9.19K      0      0      0      0      0      0
delta       15.3G   512P    148  14.8K   362K  28.4M     17     28  17.4M  28.6M      0      1    844  5.93K      0      0      0      0      0      0
delta       15.3G   512P  3.87K  14.8K  10.1M  8.09M     35      8  33.2M  8.32M      0      1    843  3.52K      0      0      0      0      0      0
delta       15.4G   512P  9.07K  1.52K  28.2M  1.12M     55      1  53.5M  1.15M      0      1    845  3.15K      0      0      0      0      0      0
delta       15.4G   512P  7.75K    131  21.5M   430K     59      1  58.3M   432K      0      1  1.15K  3.49K      0      0      0      0     10      0
delta       15.4G   512P  8.05K    138  32.2M   453K     55      1  53.5M   455K      0      1    841  3.58K      0      0      0      0      0      0
delta       15.4G   512P  7.05K    195  28.9M   647K     50      1  48.4M   650K      0      1  1.81K  3.37K      0      0      0      0      9      0
delta       15.4G   512P  7.45K    173  32.1M   574K     57      1  56.4M   577K      0      1    843  3.32K      0      0      0      0      0      0
delta       15.4G   512P  2.24K    270  7.68M   909K     44      1  42.3M   914K      0      1    843  3.77K      0      0      0      0      0      0
delta       15.4G   512P    117    182  1.64M   560K     40      1  38.7M   563K      0      1    842  3.82K      0      0      0      0      0      0
delta       15.4G   512P     38    146   383K   415K     21      0  21.2M   418K      0      0      0    188      0      0      0      0      0      0
delta       15.4G   512P     14      5  15.1K  53.3K      8      0  8.63M  53.4K      0      2      0  3.50M      0      0      0      0      0      0
delta       15.4G   512P      8  19.6K  9.13K  69.7M      6     71  6.35M  71.6M      4      1  1.45M   292K      0      0   834K     48      0      0
delta       15.7G   512P     15  20.9K  16.1K  74.1M     10     72  10.1M  72.1M      7      4  9.11M  4.99M     10      1  11.1M   451K    112      0
delta       15.1G   512P     19  15.3K  28.4K  43.3M     15     44  14.4M  44.4M      0      2  2.57K  1.93M     93     26  93.7M  11.0M    165      0
delta       15.1G   512P     24  19.6K  38.9K  22.9M     13     25  11.8M  25.6M      0      1  1.30K  2.99K      0      0      0      0      3      0
delta       15.3G   512P     62  24.8K   105K  30.1M     12     28  11.2M  27.2M      0      3  2.66K   160K      0      0      0      0      0      0
delta       15.4G   512P     31  14.9K   313K  18.5M     17     19  16.8M  18.8M      0      1    844  4.42K      0      0      0      0      2      0
delta       15.4G   512P     73  7.58K   926K  4.58M     28      5  25.8M  4.70M      0      1  1.39K  3.15K      0      0      0      0      1      0
delta       15.4G   512P     74    250  1.17M   948K     32      1  30.7M   952K      0      1    843  3.14K      0      0      0      0      0      0
delta       15.5G   512P  4.87K    338  16.3M  1.31M     39      1  37.7M  1.31M      0      1    842  3.57K      0      0      0      0      0      0
delta       15.5G   512P  6.17K    158  25.4M   626K     51      1  49.8M   628K      0      1    842  2.84K      0      0      0      0      0      0
```

**`zpool iostat -ol <pool>`**
```
                          ---------- object_agent_total ----------  ---------- object_store_total ----------
              capacity     operations    throughput      latency     operations    throughput      latency
pool        alloc   free   read  write   read  write   read  write    get    put    get    put    get    put
----------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
delta       15.6G   512P    245  3.65K   679K  8.71M   81ms  248ms      3      9  2.90M  8.93M      -      -
delta       15.6G   512P     19    140  36.0K   473K   18ms  181ms      6      3  5.95M  3.97M      -      -
delta       15.6G   512P     20  25.2K  21.3K  99.4M  146ms  158ms     85    127  72.0M   113M      -      -
delta       15.7G   512P  6.12K  14.7K  25.6M  41.1M  103ms  220ms     60     48  57.4M  49.0M      -      -
delta       15.9G   512P  8.21K  9.58K  25.2M  37.5M  136ms  180ms     94     50  92.8M  41.1M      -      -
delta       15.4G   512P    753  22.7K  2.81M  72.3M   61ms  171ms     22     77  20.1M  76.4M      -      -
delta       15.6G   512P     14  38.2K  34.9K  37.6M   91ms  534ms      4     37  4.20M  35.7M      -      -
delta       15.8G   512P     33  15.4K  40.7K  28.9M   43ms  245ms     17     32  15.0M  28.7M      -      -
delta       15.8G   512P     19  12.2K  20.1K  7.05M   44ms  260ms     12      8  10.8M  8.02M      -      -
```

ahrens commented:
It looks like there may be an additional issue, though I'm not sure if this predates your changes.  When we miss one of the scheduled printouts, the stats are wrong, they seem to be thinking that all the ops in the gap happened in one scheduled interval, rather than in the actual interval, so the "per second" values are actually :
```
$ zpool iostat -Td -ol test 1
                          ---------- object_agent_total ----------  ---------- object_store_total ----------
              capacity     operations    throughput      latency     operations    throughput      latency
pool        alloc   free   read  write   read  write   read  write    get    put    get    put    get    put
----------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
...
Wed Oct  6 04:27:54 UTC 2021
test        49.6G   512P      0  3.02K      0   148M      -  204ms    646    729   411M   526M      -      -
Wed Oct  6 04:27:58 UTC 2021
test        49.6G   512P      0  1.22K      0  59.1M      -  208ms  3.89K  3.47K  2.74G  2.29G      -      -
```
(we never achieved >1GB/s, according to `dstat`)

ahrens commented:
```
   deletions
   ops   rate
 -----  -----
```
`ops` could be unclear since we do a smaller number of calls to s3 (one call deletes many objects).
`rate` could be unclear since all of the numbers here are rates (per second).

Maybe this should be `objs  bytes` rather than `ops  rate`.

ahrens commented on cmd/zfs_object_agent/zettaobject/src/object_access.rs:
```diff
@@ -569,4 +625,60 @@ impl ObjectAccess {
+        order: Ordering,
+        s1: ObjectAccessStatType,
+        s2: ObjectAccessStatType,
+        s3: ObjectAccessStatType,
```
It seems very specific to have this function that sums 3 stat types.  If we really need that, we could call it `sum_stats3`, but I think we can make this more general:

```rust
    fn sum_stats(&self, stat_types: &[ObjectAccessStatType]) -> HashMap<String, u64> {
        let mut total = HashMap::new();

        total.insert(
            "operations".into(),
            stat_types
                .iter()
                .map(|&stat_type| {
                    self.access_stats[stat_type]
                        .operations
                        .load(Ordering::Relaxed)
                })
                .sum(),
        );
        total.insert(
            "total_bytes".into(),
            stat_types
                .iter()
                .map(|&stat_type| {
                    self.access_stats[stat_type]
                        .total_bytes
                        .load(Ordering::Relaxed)
                })
                .sum(),
        );

        total
    }
```

And then it's invoked like this:
```rust
        outer.insert(
            "TotalGet".into(),
            self.sum_stats2(&[
                ObjectAccessStatType::ReadsGet,
                ObjectAccessStatType::MetadataGet,
                ObjectAccessStatType::ReclaimGet,
            ]),
        );
```
It could be made even more general by making it take a generic `I: IntoIterator<Item=ObjectAccessStatType>`, but that probably isn't really necessary.

ahrens commented on man/man8/zpool-iostat.8:
```diff
@@ -166,6 +166,19 @@ This can be used to look up the current block device name regardless of the
+.It Sy object_store_data
+dataset blocks get/put requests to object store.
+.It Sy object_store_metadata
+pool metadata get/put request to object store.
```
how about: `agent metadata...` or `object store specific metadata ...` (since this is not the kernel's pool metadata, e.g. MOS)

ahrens commented on module/zfs/vdev.c:
```diff
@@ -4665,12 +4665,25 @@ vdev_stat_update(zio_t *zio, uint64_t psize)
+				/* object store pools don't use vdev queues */
+				if (vdev_is_object_based(vd)) {
+					vsx->vsx_disk_histo[type]
+					    [L_HISTO(zio->io_delay)]++;
```
seems like we could also track `total_histo` (which will be the same as disk_histo, which is really like "agent histo"). You could argue that we should be tracking only total_histo and leaving disk_histo blank since we don't really know it, but that may result in a worse user experience.

ahrens commented on module/zfs/vdev.c:
```diff
@@ -4665,12 +4665,25 @@ vdev_stat_update(zio_t *zio, uint64_t psize)
+				hrtime_t latency = zio->io_delta -zio->io_delay;
+
+				/* object store pools don't use vdev queues */
+				if (vdev_is_object_based(vd)) {
```
I wonder if we should be addressing this problem by setting io_delta for object store zio's?  It looks like it normally gets set in vdev_queue_io_done() which is called from zio_vdev_io_done().  And it looks like io_delay get set from zio_vdev_io_done(), so I'm guessing we go through that for object store zio's.  I'm not sure exactly what prevents us from calling vdev_queue_io_done() for object store zio's (maybe io_vd is not set???)

grwilson replied:
I think `io_delta` should always get set. If we address `io_delta`, does this mean this can just be common code?

don-brady commented:
The latest commit addresses the following:
- Added timestamp to object agent stats payload and use that when scaling the values in zpool iostat.
- The `object_store_backend_stats()` is now only called directly from `spa_config_add_object_store_stats()` rather than as a side-effect of `spa_open_common()`.
- Fixed the `vsx_total_histo` stat and corresponding `zio->io_delta` (it is now computed in `vdev_object_store`)
- Suppress the `vsx_disk_histo` stat for object based pools.
- Added a reissue for pending stat callers in `agent_resume()`.
- Other cleanup from review feedback.

ahrens commented on cmd/zfs_object_agent/zettaobject/src/object_access.rs:
```diff
@@ -546,6 +617,10 @@ impl ObjectAccess {
                 })
                 .await
                 .unwrap();
+                // XXX need to also update the object sizes
```
you can remove the XXX since we decided not to do that.

grwilson commented on module/os/linux/zfs/vdev_object_store.c:
```diff
@@ -1551,6 +1676,7 @@ vdev_object_store_io_start(zio_t *zio)
 static void
 vdev_object_store_io_done(zio_t *zio)
 {
+	zio->io_delta = gethrtime() - zio->io_timestamp;
```
This should not be necessary since this should be set by `vdev_queue_io_done`.

grwilson commented on module/zfs/spa_config.c:
```diff
@@ -541,6 +542,39 @@ spa_config_generate(spa_t *spa, vdev_t *vd, uint64_t txg, int getstats)
+
+	ASSERT(spa_is_object_based(spa));
+
+	spa_config_enter(spa, SCL_CONFIG | SCL_STATE, FTAG, RW_READER);
```
Can you comment why the `SCL_STATE` lock is required. We already have the `SCL_CONFIG` lock (which we can assert)

grwilson commented on module/zfs/spa_config.c:
```diff
@@ -541,6 +542,39 @@ spa_config_generate(spa_t *spa, vdev_t *vd, uint64_t txg, int getstats)
 	return (config);
 }
 
+/*
+ * Add the object store specific stats to an existing config
+ */
+void
+spa_config_add_object_store_stats(spa_t *spa, nvlist_t *config)
+{
+	vdev_t *rvd;
+	nvlist_t *nvl, *nvroot, **child;
```
I think a better way to accomplish this is to define a `vdev_op_config_generate` for `vdev_object_store` vdevs. This should simplify the logic and allow you to add the to the nvlist at the vdev level.

don-brady commented on module/os/linux/zfs/vdev_object_store.c:
```diff
@@ -932,6 +966,57 @@ object_store_get_stats(vdev_t *vd, vdev_object_store_stats_t *vossp)
+
+	stats_call.oss_nvl = NULL;
+	stats_call.oss_owner = (uint64_t)curthread;
+	mutex_init(&stats_call.oss_lock, NULL, MUTEX_DEFAULT, NULL);
+	cv_init(&stats_call.oss_cv, NULL, CV_DEFAULT, NULL);
+
```
adding missing destroy calls
